engine/jobmaster/dm/openapi.go

```
@@ -0,0 +1,251 @@
+func (jm *JobMaster) DMAPIGetBinlogOperator(c *gin.Context, taskName string, params openapi.DMAPIGetBinlogOperatorParams) {
+	req := &dmpkg.BinlogRequest{
+		Op:      pb.ErrorOp_List,
+		Sources: []string{taskName},
+	}
+	if params.BinlogPos != nil {
+		req.BinlogPos = *params.BinlogPos
+	}
+	resp, err := jm.Binlog(c.Request.Context(), req)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, resp)
+}
+
+// DMAPISetBinlogOperator implements the api of set binlog operator.
+func (jm *JobMaster) DMAPISetBinlogOperator(c *gin.Context, taskName string) {
+	var req openapi.SetBinlogOperatorRequest
+	if err := c.Bind(&req); err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	r := &dmpkg.BinlogRequest{
+		Sources: []string{taskName},
+	}
+	if req.BinlogPos != nil {
+		r.BinlogPos = *req.BinlogPos
+	}
+	if req.Sqls != nil {
+		r.Sqls = *req.Sqls
+	}
+	switch req.Op {
+	case openapi.SetBinlogOperatorRequestOpInject:
+		r.Op = pb.ErrorOp_Inject
+	case openapi.SetBinlogOperatorRequestOpSkip:
+		r.Op = pb.ErrorOp_Skip
+	case openapi.SetBinlogOperatorRequestOpReplace:
+		r.Op = pb.ErrorOp_Replace
+	default:
+		_ = c.Error(errors.New("unsupport op type '' for set binlog operator"))
+		return
+	}
+	resp, err := jm.Binlog(c.Request.Context(), r)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, resp)
+}
+
+// DMAPIDeleteBinlogOperator implements the api of delete binlog operator.
+func (jm *JobMaster) DMAPIDeleteBinlogOperator(c *gin.Context, taskName string, params openapi.DMAPIDeleteBinlogOperatorParams) {
+	req := &dmpkg.BinlogRequest{
+		Op:      pb.ErrorOp_Revert,
+		Sources: []string{taskName},
+	}
+	if params.BinlogPos != nil {
+		req.BinlogPos = *params.BinlogPos
+	}
+	resp, err := jm.Binlog(c.Request.Context(), req)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	// TODO: use no content response, now return content because delete may failed.
```

Not sure if we should return a 500 errors when resp.ErrorMsg is not empty